discover/discover.go

```diff
@@ -0,0 +1,85 @@
+package discover
+
+import (
+	"fmt"
+	"log"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/discover/aws"
+	"github.com/hashicorp/discover/azure"
+	"github.com/hashicorp/discover/gce"
+)
+
+func Discover(cfg string, l *log.Logger) ([]string, error) {
```

Comment:
I think interfaces should be defined with the consumer and not the producer. If someone needs a `Discoverer` interface then they can easily define it. 

Another issue is that although the signature of the `(aws|gce|azure).Discover` methods are similar they differ in their config structures. Therefore, the `discover.Discover` method is the generalized function. We might want to consider passing either a string and provide a parser or to pass a `map[string]string` to make the signature truly generic.